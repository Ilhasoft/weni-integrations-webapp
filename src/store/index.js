import Vue from 'vue';
import Vuex from 'vuex';

import Auth from './auth';

Vue.use(Vuex);

const store = new Vuex.Store({
  modules: {
    Auth,
  },
});

store.dispatch('retriveAuthToken');
store.dispatch('retriveSelectedOrg');
store.dispatch('retriveSelectedProject');

export default store;
