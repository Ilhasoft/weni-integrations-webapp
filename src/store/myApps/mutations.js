export default {
  GET_CONFIGURED_APPS_REQUEST(state, skipLoading = false) {
    state.loadingConfiguredApps = !skipLoading;
    state.errorConfiguredApps = null;

    if (!skipLoading) {
      state.configuredApps = null;
    }
  },
  GET_CONFIGURED_APPS_SUCCESS(state, data) {
    state.configuredApps = data;
    state.loadingConfiguredApps = false;
  },
  GET_CONFIGURED_APPS_ERROR(state, err) {
    state.errorConfiguredApps = err;
    state.loadingConfiguredApps = false;
  },

  GET_INSTALLED_APPS_REQUEST(state) {
    state.loadingInstalledApps = true;
    state.errorInstalledApps = null;
    state.installedApps = null;
  },
  GET_INSTALLED_APPS_SUCCESS(state, data) {
    state.installedApps = data;
    state.loadingInstalledApps = false;
  },
  GET_INSTALLED_APPS_ERROR(state, err) {
    state.errorInstalledApps = err;
    state.loadingInstalledApps = false;
  },
};
