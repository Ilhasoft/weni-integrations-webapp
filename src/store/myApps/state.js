export default {
  configuredApps: [],
  loadingConfiguredApps: false,
  errorConfiguredApps: null,

  installedApps: [],
  loadingInstalledApps: false,
  errorInstalledApps: null,
};
