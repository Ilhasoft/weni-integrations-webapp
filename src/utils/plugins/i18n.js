import Vue from 'vue';
import VueI18n from 'vue-i18n';

Vue.use(VueI18n);

import en from '../../locales/en.json';
import pt_br from '../../locales/pt_br.json';
import es_es from '../../locales/es_es.json';

const languages = {
  'en-us': en,
  'pt-br': pt_br,
  es: es_es,
};

const messages = Object.assign(languages);

const i18n = new VueI18n({
  locale: 'en-us',
  fallbackLocale: 'en-us',
  messages,
});

export default i18n;
