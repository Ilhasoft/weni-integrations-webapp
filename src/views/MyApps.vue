<template>
  <div>
    <app-grid section="configured" type="edit" />
    <app-grid section="installed" type="config" />
  </div>
</template>

<script>
  import AppGrid from '../components/AppGrid.vue';
  export default {
    name: 'Apps',
    components: { AppGrid },
  };
</script>

<style scoped></style>
