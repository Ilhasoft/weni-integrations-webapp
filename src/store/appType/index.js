import state from './state';
import actions from './actions';
import mutations from './mutations';

import channels from './channels';
import comments from './comments';
import externals from './externals';

export default {
  modules: {
    channels,
    comments,
    externals,
  },
  actions,
  mutations,
  state,
};
