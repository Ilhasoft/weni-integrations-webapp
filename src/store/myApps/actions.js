import appType from '@/api/myApps';
import { clearHtmlTags } from '@/utils/clearHtmlTags';

/* istanbul ignore next */
function formatGenericApp(appList) {
  return appList.map((app) => {
    const summary =
      app.code === 'generic' ? clearHtmlTags(app.config.channel_claim_blurb) : app.summary;

    return { ...app, generic: app.code === 'generic', summary };
  });
}

export default {
  async getConfiguredApps({ commit }, { params }) {
    commit('GET_CONFIGURED_APPS_REQUEST');
    try {
      let { data } = await appType.getConfiguredApps(params);

      data = formatGenericApp(data);

      commit('GET_CONFIGURED_APPS_SUCCESS', data);
    } catch (err) {
      commit('GET_CONFIGURED_APPS_ERROR', err);
    }
  },
  async getInstalledApps({ commit }, { params }) {
    commit('GET_INSTALLED_APPS_REQUEST');
    try {
      let { data } = await appType.getInstalledApps(params);

      data = formatGenericApp(data);

      commit('GET_INSTALLED_APPS_SUCCESS', data);
    } catch (err) {
      commit('GET_INSTALLED_APPS_ERROR', err);
    }
  },
};
