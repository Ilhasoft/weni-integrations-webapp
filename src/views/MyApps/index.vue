<template>
  <div class="my-apps">
    <unnnic-autocomplete
      v-if="hasApps"
      v-model="searchTerm"
      class="my-apps__search"
      :placeholder="$t('apps.discovery.search.placeholder')"
      icon-left="search-1"
      :data="searchOptions"
      id="search"
    />

    <span
      v-if="hasApps && searchTerm && searchTerm.trim()"
      class="discovery-content__search__results"
    >
      {{ $t('apps.discovery.search.results') }}
      <span class="discovery-content__search__results__highlight">
        {{ `“${searchTerm}”...` }}
      </span>
    </span>

    <div v-if="hasApps" class="my-apps__sections">
      <div v-if="searchTerm && !filteredConfiguredApps.length && !filteredInstalledApps.length">
        <EmptyApps :term="searchTerm" @clear="navigateToDiscovery" />
      </div>
      <div class="my-apps__sections__grids" v-else>
        <app-grid
          section="configured"
          type="edit"
          :loading="loadingConfiguredApps"
          :apps="filteredConfiguredApps"
          @update="fetchConfigured"
        />
        <app-grid
          section="installed"
          type="config"
          :loading="loadingInstalledApps"
          :apps="filteredInstalledApps"
          @update="fetchInstalled"
        />
      </div>
    </div>
    <div v-else class="my-apps__empty">
      <img class="my-apps__empty__image" src="@/assets/svgs/empty-apps.svg" />
      <div class="my-apps__empty__description__main">
        {{ $t('apps.myApps.empty.no-integrations') }}
      </div>
      <div class="my-apps__empty__description__secondary">
        <div>{{ $t('apps.myApps.empty.search') }}</div>
        <div class="my-apps__empty__description__secondary__link" @click="navigateToDiscovery()">
          {{ $t('apps.myApps.empty.link') }}
        </div>
      </div>
    </div>
  </div>
</template>

<script>
  import AppGrid from '@/components/AppGrid/index.vue';
  import EmptyApps from '@/components/EmptyApps/index.vue';
  import { mapState, mapActions } from 'pinia';
  import unnnic from '@weni/unnnic-system';
  import { auth_store } from '@/stores/modules/auth.store';
  import { my_apps } from '@/stores/modules/myApps.store';
  import { useEventStore } from '@/stores/event.store';

  export default {
    name: 'Apps',
    components: { AppGrid, EmptyApps },
    data() {
      return {
        searchTerm: '',
        eventStore: useEventStore(),
      };
    },
    /* istanbul ignore next */
    mounted() {
      this.fetchCategories();
      this.on('updateGrid', this.fetchCategories);
    },
    beforeUnmount() {
      this.off('updateGrid', this.fetchCategories);
    },
    /* istanbul ignore next */
    computed: {
      ...mapState(auth_store, ['project']),
      ...mapState(my_apps, [
        'configuredApps',
        'errorConfiguredApps',
        'errorConfiguredApps',
        'installedApps',
        'loadingInstalledApps',
        'errorInstalledApps',
      ]),
      hasApps: function () {
        if (this.loadingConfiguredApps || this.loadingInstalledApps) {
          return true;
        }

        return this.configuredApps?.length || this.installedApps?.length;
      },
      searchOptions() {
        if (!this.configuredApps || !this.installedApps) return [];

        const allApps = [...this.configuredApps, ...this.installedApps];

        const filtered = this.filterByName(allApps, this.searchTerm);

        return filtered.map((app) => {
          let name = app.generic ? app.config.channel_name : app.name;
          if (app.config?.title) {
            name += ` - ${app.config?.title}`;
          }
          return name;
        });
      },
      filteredConfiguredApps() {
        if (!this.configuredApps) return [];

        if (!this.searchTerm || !this.searchTerm.trim()) return this.configuredApps;

        return this.filterByName(this.configuredApps, this.searchTerm);
      },
      filteredInstalledApps() {
        if (!this.installedApps) return [];

        if (!this.searchTerm || !this.searchTerm.trim()) return this.installedApps;

        return this.filterByName(this.installedApps, this.searchTerm);
      },
    },
    methods: {
      ...mapActions(my_apps, ['getConfiguredApps', 'getInstalledApps']),
      ...mapActions(useEventStore, ['on', 'off']),
      filterByName(appList, search) {
        return appList.filter((app) => {
          const appMainName = app.generic ? app.config.channel_name : app.name;
          const appName = app.config?.title || app.config?.channel_name;

          const name = appMainName + appName;
          return name.toLowerCase().includes(search.toLowerCase());
        });
      },
      fetchCategories() {
        this.fetchConfigured();
        this.fetchInstalled();
      },
      fetchConfigured() {
        const params = {
          project_uuid: this.project,
        };
        this.getConfiguredApps({ params });

        if (this.errorConfiguredApps) {
          this.callErrorModal({ text: this.$t('apps.myApps.error.configured') });
        }
      },
      fetchInstalled() {
        const params = {
          project_uuid: this.project,
        };
        this.getInstalledApps({ params });

        if (this.errorInstalledApps) {
          this.callErrorModal({ text: this.$t('apps.myApps.error.installed') });
        }
      },
      navigateToDiscovery() {
        this.$router.replace('/apps/discovery');
      },
      callErrorModal({ text }) {
        unnnic.unnnicCallAlert({
          props: {
            text: text,
            type: 'error',
          },
          seconds: 6,
        });
      },
      handleEvent() {
        console.log('chegou');
        this.fetchCategories();
      },
    },
  };
</script>

<style lang="scss" scoped>
  @import './styles.scss';
</style>
