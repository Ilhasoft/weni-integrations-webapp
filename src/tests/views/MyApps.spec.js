import { mount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import MyApps from '../../views/MyApps/index.vue';
import { describe, expect, it, vi, beforeEach } from 'vitest';
import i18n from '@/utils/plugins/i18n';
import { defineStore, setActivePinia, createPinia } from 'pinia';
import { my_apps } from '@/stores/modules/myApps.store';
import { wrap } from '@sentry/vue';

describe('MyApps', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });
  it('matches snapshot', () => {
    const wrapper = mount(MyApps, {
      global: {
        plugins: [createTestingPinia(), i18n],
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('mounted', () => {
    it('should call fetchCategories', async () => {
      const spy = vi.spyOn(MyApps.methods, 'fetchCategories');
      const wrapper = mount(MyApps, {
        global: {
          plugins: [createTestingPinia({ createSpy: vi.fn }), i18n],
        },
      });
      await wrapper.vm.$nextTick();
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });

  describe('methods', () => {
    it('should call fetchInstalled', async () => {
      const fetchInstalledMock = vi
        .spyOn(MyApps.methods, 'fetchInstalled')
        .mockImplementation(() => {});
      const fetchConfiguredMock = vi
        .spyOn(MyApps.methods, 'fetchConfigured')
        .mockImplementation(() => {});

      const wrapper = mount(MyApps, {
        global: {
          plugins: [createTestingPinia({ createSpy: vi.fn }), i18n],
        },
      });

      wrapper.vm.fetchCategories();
      expect(fetchInstalledMock).toHaveBeenCalledTimes(2);

      fetchInstalledMock.mockRestore();
      fetchConfiguredMock.mockRestore();
    });

    it('should call getConfiguredApps', async () => {
      const getConfiguredApps = vi
        .spyOn(MyApps.methods, 'getConfiguredApps')
        .mockImplementation(() => {});

      const wrapper = mount(MyApps, {
        global: {
          plugins: [createTestingPinia({ createSpy: vi.fn }), i18n],
        },
      });

      wrapper.vm.fetchConfigured();
      expect(getConfiguredApps).toHaveBeenCalledTimes(2);

      getConfiguredApps.mockRestore();
    });

    it('should call getInstalledApps', async () => {
      const getInstalledApps = vi
        .spyOn(MyApps.methods, 'getInstalledApps')
        .mockImplementation(() => {});

      const wrapper = mount(MyApps, {
        global: {
          plugins: [createTestingPinia({ createSpy: vi.fn }), i18n],
        },
      });

      wrapper.vm.fetchInstalled();
      expect(getInstalledApps).toHaveBeenCalledTimes(2);

      getInstalledApps.mockRestore();
    });

    it('should change route', async () => {
      const replaceMock = vi.fn();
      const wrapper = mount(MyApps, {
        global: {
          plugins: [createTestingPinia({ createSpy: vi.fn }), i18n],
          mocks: {
            $router: {
              replace: replaceMock,
            },
          },
        },
      });

      wrapper.vm.navigateToDiscovery();
      expect(replaceMock).toHaveBeenCalledWith('/apps/discovery');
    });
  });

  describe('template', () => {
    beforeEach(() => {
      vi.clearAllMocks();
    });
    it('should render the autocomplete component with the correct props', async () => {
      const pinia = createTestingPinia({
        stubActions: false,
        createSpy: vi.fn,
      });
      const wrapper = mount(MyApps, {
        global: {
          plugins: [pinia, i18n],
        },
      });
      await wrapper.vm.$nextTick();
      const autocomplete = wrapper.find('#search');
      expect(wrapper.vm.hasApps).toBe(true);
      expect(autocomplete.exists()).toBe(true);
    });
  });
});
