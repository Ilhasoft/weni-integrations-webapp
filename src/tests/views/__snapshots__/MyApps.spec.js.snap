// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MyApps > matches snapshot 1`] = `
"<div data-v-d693ca5e="" class="my-apps">
  <!--v-if-->
  <!--v-if-->
  <div data-v-d693ca5e="" class="my-apps__empty"><img data-v-d693ca5e="" class="my-apps__empty__image" src="/src/assets/svgs/empty-apps.svg">
    <div data-v-d693ca5e="" class="my-apps__empty__description__main">You haven't installed any integrations yet.</div>
    <div data-v-d693ca5e="" class="my-apps__empty__description__secondary">
      <div data-v-d693ca5e="">Search for what you need</div>
      <div data-v-d693ca5e="" class="my-apps__empty__description__secondary__link">here</div>
    </div>
  </div>
</div>"
`;
